sympy/ntheory/residue_ntheory.py

```diff
@@ -70,26 +69,21 @@ def n_order(a, n):
         return 1
     if gcd(a, n) != 1:
         raise ValueError("The two numbers should be relatively prime")
-    # We want to calculate
-    # order = totient(n), factors = factorint(order)
-    factors = defaultdict(int)
-    for px, kx in factorint(n).items():
-        if kx > 1:
-            factors[px] += kx - 1
-        for py, ky in factorint(px - 1).items():
-            factors[py] += ky
-    order = 1
-    for px, kx in factors.items():
-        order *= px**kx
-    # Now the `order` is the order of the group.
-    # The order of `a` divides the order of the group.
-    for p, e in factors.items():
-        for _ in range(e):
-            if pow(a, order // p, n) == 1:
-                order //= p
-            else:
-                break
-    return order
+    a_order = 1
+    for p, e in factorint(n).items():
+        pe = p**e
+        pe_order = (p - 1) * p**(e - 1)
+        factors = factorint(p - 1)
+        if e > 1:
+            factors[p] = e - 1
+        order = 1
+        for px, ex in factors.items():
+            x = pow(a, pe_order // px**ex, pe)
+            while x != 1:
```

Comment:
Isn't this an `if`?

Reply:
The current version decrements from the larger `px` exponent and looks for the first exponent that is not 1.
The PR version increments from the smallest `px` exponent and looks for the first exponent that is 1.
So this is correct.

Reply:
I see that (a,n) = (23,97) needs 5 iterations and if you don't use `lcm` then (a,n)=(17,76) gives 1 instead of 9.

Reply:
Is this a reply to my other comment? LCM is required for order if n is factored into prime powers, but I didn't benchmark if this is any faster than running the algorithm on n directly.